gnucash/import-export/import-main-matcher.cpp

```diff
@@ -192,32 +195,95 @@ defer_bal_computation (GNCImportMainMatcher *info, Account* acc)
     }
 }
 
+
+// Custom iterator class
+class GtkTreeModelIterator
+{
+public:
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type = ptrdiff_t;
+    GtkTreeModelIterator(GtkTreeModel* model, std::optional<GtkTreeIter> iter)
+        : m_model(model), m_iter(iter) {}
+    GtkTreeModelIterator& operator++()
+    {
+        if (m_iter.has_value() && !gtk_tree_model_iter_next(m_model, &*m_iter))
+            m_iter = std::nullopt;
+        return *this;
+    }
+    template <typename T>
+    T get (int column) {
+        g_return_val_if_fail (m_iter.has_value(), T());
+        gpointer rv;
+        gtk_tree_model_get(m_model, &*m_iter, column, &rv, -1);
+        return static_cast<T>(rv);
+    }
+    GNCImportTransInfo* get_trans_info ()
+    {
+        g_return_val_if_fail (m_iter.has_value(), nullptr);
+        return get<GNCImportTransInfo*>(DOWNLOADED_COL_DATA);
+    };
+    bool has_value() { return m_iter.has_value(); };
+    std::string get_account_name ()
+    {
+        g_return_val_if_fail (m_iter.has_value(), "");
+        auto str = get<char*>(DOWNLOADED_COL_ACCOUNT);
+        std::string rv{str};
+        g_free (str);
+        return rv;
+    }
+    GtkTreeModelIterator operator*() const { return *this; }
+    bool operator==(const GtkTreeModelIterator& other) const
+    {
+        if (!m_iter.has_value() && !other.m_iter.has_value())
+            return true;
+        if (!m_iter.has_value() || !other.m_iter.has_value())
+            return false;
+        return (m_model == other.m_model) &&
+               (m_iter->stamp == other.m_iter->stamp) &&
+               (m_iter->user_data == other.m_iter->user_data) &&
+               (m_iter->user_data2 == other.m_iter->user_data2) &&
+               (m_iter->user_data3 == other.m_iter->user_data3);
+    }
+    bool operator!=(const GtkTreeModelIterator& other) const { return !(*this == other); }
+private:
+    GtkTreeModel* m_model;
+    std::optional<GtkTreeIter> m_iter;
+};
+
+// Custom container class
+class GtkTreeModelContainer
+{
+public:
+    using iterator = GtkTreeModelIterator;
+    GtkTreeModelContainer(GtkTreeModel* model) : m_model(model), m_size{0}
+    {
+        for (auto it = begin(); it.has_value(); ++it)
+            m_size++;
+    }
```

Comment:
this O(N) to calculate `size` is inevitable. Alternately we could forego size() and test empty via `return begin() == end()`...?

Reply:
Constructing `end()` is trivial, so that would be much faster for `empty()`. This will anyway work to get the size only of a GtkListModel because a GtkTreeModel's iterators have children reflecting its tree structure.

Reply:
> This will anyway work to get the size only of a GtkListModel because a GtkTreeModel's iterators have children reflecting its tree structure.

Not quite sure what this means.